server/src/services/database.service.spec.ts

```diff
@@ -9,287 +11,254 @@ describe(DatabaseService.name, () => {
   let sut: DatabaseService;
   let databaseMock: Mocked<IDatabaseRepository>;
   let loggerMock: Mocked<ILoggerRepository>;
+  let versionRange: string;
+  let versionBelowRange: string;
+  let minVersionInRange: string;
+  let patchUpgradeInRange: string;
+  let versionAboveRange: string;
+
+  describe.each([
+    { extension: DatabaseExtension.VECTOR, extensionName: EXTENSION_NAMES[DatabaseExtension.VECTOR] },
+    { extension: DatabaseExtension.VECTORS, extensionName: EXTENSION_NAMES[DatabaseExtension.VECTORS] },
+  ])('should work with $extensionName', ({ extension, extensionName }) => {
+    beforeEach(() => {
+      process.env.DB_VECTOR_EXTENSION = extensionName;
+      databaseMock = newDatabaseRepositoryMock();
+      loggerMock = newLoggerRepositoryMock();
+      sut = new DatabaseService(databaseMock, loggerMock);
+
+      let minVersion;
+      if (extension === DatabaseExtension.VECTOR) {
+        minVersion = semver.minVersion(VECTOR_VERSION_RANGE)!;
+        versionRange = VECTOR_VERSION_RANGE;
+      } else {
+        minVersion = semver.minVersion(VECTORS_VERSION_RANGE)!;
+        versionRange = VECTORS_VERSION_RANGE;
+      }
+      versionBelowRange = `${minVersion.major}.${minVersion.minor - 1}.${minVersion.patch}`;
+      minVersionInRange = minVersion.version;
+      patchUpgradeInRange = minVersion.inc('patch')!.version;
+      versionAboveRange = minVersion.inc('major')!.version;
+      databaseMock.getExtensionVersion.mockResolvedValue({
+        installedVersion: minVersionInRange,
+        availableVersion: minVersionInRange,
+      });
+    });
```

Comment:
This defeats the whole point of the unit test in my opinion. You aren't actually testing with specific values you are programmatically writing tests to test the code. You could just as easily introduce a bug in _this_ code. Tests should have less programming and more static assertions otherwise they lose a lot of value. I previously changed the tests to be less like that and now you have just reverted everything again. 

Like we should not be using semver to generate tests for our code that _relies_ on semver.

Reply:
The existing unit tests are a huge pain honestly. They're brittle because they stop testing the same thing as soon as a version constraint changes, they're repetitive because you have the same test for both extensions, and there were several mistakes in the unit tests that mixed up the two because they're hard to spot. It's just all very manual and error-prone.

Reply:
Maybe we need to move getting the expected range to a repository call, because then the tests could validate everything without relying on a specific version that will change over time. Programming the brittleness away also isn't the answer.

Reply:
Unless we have logic that is different/specific for a particular extension we don't need to be unit trying everything twice.

Reply:
> Maybe we need to move getting the expected range to a repository call, because then the tests could validate everything without relying on a specific version that will change over time.

I can add a repo method for that, mock it in the `beforeEach` with a specific range and set fixed values for the other variables. Does that sound good?

> Unless we have logic that is different/specific for a particular extension we don't need to be unit testing everything twice.

I think it's easier and less error-prone to just run them on both. I originally tried working with the existing unit tests, but going back and forth to see if this behavior is being tested for this extension is no fun. It's also easy to forget to add it for one, make a change in one test and not apply it to the other, etc.

Reply:
Mocking the version and using static variables for the rest is fine. 

Testing both extensions all the time is also fine. The problem for me is when your test cases start having stuff like "if the extension is this, then do this one off assertions". Basically it is the same problem again. 

You could run a bunch of tests for both with the same assertions, but if there is some specific logic for a specific scenario that should be in a dedicated test imo (without if statements).

Reply:
That makes sense. I think the "suggest the other extension" check falls in that category, so I can move that out into two dedicated tests.

Reply:
They would be perfect, thanks.